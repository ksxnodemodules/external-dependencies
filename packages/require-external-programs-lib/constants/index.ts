export const KEYWORD = 'require-external-programs'
export const KEYNAME = 'requireExternalPrograms'
export const VERSION = '0.0.7-rc'
