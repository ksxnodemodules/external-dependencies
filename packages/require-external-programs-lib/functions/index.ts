import allManifests from './all-manifests'
import fromManifests from './from-manifests'
import classifyManifestItem from './classify-manifest-item'
import classifyManifestList from './classify-manifest-list'
import fromAllManifests from './from-all-manifests'

export {
  allManifests,
  fromManifests,
  classifyManifestItem,
  classifyManifestList,
  fromAllManifests
}
