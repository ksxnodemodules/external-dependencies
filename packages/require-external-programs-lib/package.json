{
  "name": "require-external-programs-lib",
  "description": "Programmatic library for require-external-programs package",
  "version": "0.0.6",
  "author": "Hoàng Văn Khải <hvksmr1996@gmail.com>",
  "license": "MIT",
  "homepage": "https://github.com/ksxnodemodules/external-dependencies#readme",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/ksxnodemodules/external-dependencies.git"
  },
  "bugs": {
    "url": "https://github.com/ksxnodemodules/external-dependencies/issues"
  },
  "keywords": [
    "lib",
    "api",
    "external",
    "programs",
    "require"
  ],
  "dependencies": {
    "semver": "^7.0.0",
    "fs-extra": "^7.0.1",
    "glob": "^7.1.3",
    "glob-promise": "^3.4.0",
    "advanced-set-mutable-multi-key": "^0.0.2",
    "tslib": "^1.9.3",
    "@types/semver": "^7.0.0",
    "@types/fs-extra": "^5.0.4",
    "@types/node": "^10.12.18"
  },
  "engines": {
    "node": ">= 8.9.0"
  }
}
