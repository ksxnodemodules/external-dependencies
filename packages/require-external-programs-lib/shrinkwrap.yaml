dependencies:
  '@types/fs-extra': 5.0.4
  '@types/node': 10.12.5
  '@types/semver': 5.5.0
  advanced-set-mutable-multi-key: 0.0.1
  fs-extra: 7.0.1
  glob: 7.1.3
  glob-promise: 3.4.0
  semver: 5.6.0
  tslib: 1.9.3
packages:
  /@babel/runtime/7.1.5:
    dependencies:
      regenerator-runtime: 0.12.1
    dev: false
    resolution:
      integrity: sha512-xKnPpXG/pvK1B90JkwwxSGii90rQGKtzcMt2gI5G6+M0REXaq6rOHsGC2ay6/d0Uje7zzvSzjEzfR3ENhFlrfA==
  /@types/events/1.2.0:
    dev: false
    resolution:
      integrity: sha512-KEIlhXnIutzKwRbQkGWb/I4HFqBuUykAdHgDED6xqwXJfONCjF5VoE0cXEiurh3XauygxzeDzgtXUqvLkxFzzA==
  /@types/fs-extra/5.0.4:
    dependencies:
      '@types/node': 10.12.5
    dev: false
    resolution:
      integrity: sha512-DsknoBvD8s+RFfSGjmERJ7ZOP1HI0UZRA3FSI+Zakhrc/Gy26YQsLI+m5V5DHxroHRJqCDLKJp7Hixn8zyaF7g==
  /@types/glob/7.1.1:
    dependencies:
      '@types/events': 1.2.0
      '@types/minimatch': 3.0.3
      '@types/node': 10.12.5
    dev: false
    resolution:
      integrity: sha512-1Bh06cbWJUHMC97acuD6UMG29nMt0Aqz1vF3guLfG+kHHJhy3AyohZFFxYk2f7Q1SQIrNwvncxAE0N/9s70F2w==
  /@types/minimatch/3.0.3:
    dev: false
    resolution:
      integrity: sha512-tHq6qdbT9U1IRSGf14CL0pUlULksvY9OZ+5eEgl1N7t+OA3tGvNpxJCzuKQlsNgCVwbAs670L1vcVQi8j9HjnA==
  /@types/node/10.12.5:
    dev: false
    resolution:
      integrity: sha512-GzdHjq3t3eGLMv92Al90Iq+EoLL+86mPfQhuglbBFO7HiLdC/rkt+zrzJJumAiBF6nsrBWhou22rPW663AAyFw==
  /@types/semver/5.5.0:
    dev: false
    resolution:
      integrity: sha512-41qEJgBH/TWgo5NFSvBCJ1qkoi3Q6ONSF2avrHq1LVEZfYpdHmj0y9SuTK+u9ZhG1sYQKBL1AWXKyLWP4RaUoQ==
  /advanced-set-altered-equal/0.0.1:
    dependencies:
      '@types/node': 10.12.5
      advanced-set-base: 0.0.0
      advanced-set-types: 0.0.0
      convenient-typescript-utilities: 0.0.3
      tslib: 1.9.3
    dev: false
    engines:
      node: '>= 8.9.0'
    resolution:
      integrity: sha512-WvpNv4/zsZc6MYXPlpY0xGnjXXmWFWOUvc43eFfqKgSBWL6n5QLbkyneyjRdF6sBjDKXe2Yla4u1VsRt8IEdEQ==
  /advanced-set-base/0.0.0:
    dependencies:
      '@types/node': 10.12.5
      advanced-set-types: 0.0.0
      tslib: 1.9.3
      typescript-miscellaneous: 0.1.0
    deprecated: Depends on wrong version of "convenient-typescript-utilities"
    dev: false
    engines:
      node: '>= 8.9.0'
    resolution:
      integrity: sha512-RP3zt5zRkNTAWCW/0u1A3jmOp6xel8qFt0U9YYbuhNRGIkjGhHOb9lzchbND6yutVohaH32Voz0ojbOVIzkZLg==
  /advanced-set-multi-key-base/0.0.1:
    dependencies:
      '@types/node': 10.12.5
      advanced-set-altered-equal: 0.0.1
      advanced-set-types: 0.0.0
      create-array-equal: 0.0.0
      tslib: 1.9.3
    dev: false
    engines:
      node: '>= 8.9.0'
    resolution:
      integrity: sha512-kCZJkQZzpCbdmn830Bv6OX1MA9FlAuo3zJB0X6gdyPMqKSiA+lLk1JuBBEBjjT+XJdM1ISGMoUg70s16xOZTPg==
  /advanced-set-mutable-multi-key/0.0.1:
    dependencies:
      '@types/node': 10.12.5
      advanced-set-multi-key-base: 0.0.1
      advanced-set-types: 0.0.0
      tslib: 1.9.3
    dev: false
    engines:
      node: '>= 8.9.0'
    resolution:
      integrity: sha512-+9XfZs67S9r5Xh3wXNYDWZaspBHmKL1YA+NOhSsQSteRMrs6d8prPto0DB2bvXExkvlbiBG9Wo6pXwhQgVe2RA==
  /advanced-set-types/0.0.0:
    dependencies:
      '@types/node': 10.12.5
      tslib: 1.9.3
      typescript-miscellaneous: 0.1.0
    deprecated: Depends on wrong version of "convenient-typescript-utilities"
    dev: false
    engines:
      node: '>= 8.9.0'
    resolution:
      integrity: sha512-xAvd7khFkgWxmedqyBb53RW9JJwEOWahX/h2LaKx6YJkjHkriBPVkW1zX6K1RabUnWza6loqoKooe/oZkEQ+OQ==
  /balanced-match/1.0.0:
    dev: false
    resolution:
      integrity: sha1-ibTRmasr7kneFk6gK4nORi1xt2c=
  /brace-expansion/1.1.11:
    dependencies:
      balanced-match: 1.0.0
      concat-map: 0.0.1
    dev: false
    resolution:
      integrity: sha512-iCuPHDFgrHX7H2vEI/5xpz07zSHB00TpugqhmYtVmMO6518mCuRMoOYFldEBl0g187ufozdaHgWKcYFb61qGiA==
  /concat-map/0.0.1:
    dev: false
    resolution:
      integrity: sha1-2Klr13/Wjfd5OnMDajug1UBdR3s=
  /convenient-typescript-utilities/0.0.3:
    dependencies:
      '@types/node': 10.12.5
      tslib: 1.9.3
    dev: false
    engines:
      node: '>= 8.9.0'
    resolution:
      integrity: sha512-kHjMz9faySaKMx9ZRwv9oTogYRXCGaKPYq+xsLEswBo+0a5r8l5Xk/7Q+vqDLMYMrTm2oW7wEgtiDcXonr8Nkg==
  /create-array-equal/0.0.0:
    dependencies:
      '@types/node': 10.12.5
      iter-tools: 6.1.6
      tslib: 1.9.3
      typescript-miscellaneous: 0.1.0
    dev: false
    engines:
      node: '>= 8.9.0'
    resolution:
      integrity: sha512-PKGYy366uoR4/A1RqKgVmyZfxf7M7Ilp9L7zHx7OubCylYp4G2WcS3gnLwrfSKIUfbV8jOOYqY1YbNOhVGhR9g==
  /dequeue/1.0.5:
    dev: false
    resolution:
      integrity: sha1-EPHO8H4yNLIdyzj0v6LWYDSrZ8c=
  /fs-extra/7.0.1:
    dependencies:
      graceful-fs: 4.1.15
      jsonfile: 4.0.0
      universalify: 0.1.2
    dev: false
    engines:
      node: '>=6 <7 || >=8'
    resolution:
      integrity: sha512-YJDaCJZEnBmcbw13fvdAM9AwNOJwOzrE4pqMqBq5nFiEqXUqHwlK4B+3pUw6JNvfSPtX05xFHtYy/1ni01eGCw==
  /fs.realpath/1.0.0:
    dev: false
    resolution:
      integrity: sha1-FQStJSMVjKpA20onh8sBQRmU6k8=
  /glob-promise/3.4.0:
    dependencies:
      '@types/glob': 7.1.1
    dev: false
    engines:
      node: '>=4'
    peerDependencies:
      glob: '*'
    resolution:
      integrity: sha512-q08RJ6O+eJn+dVanerAndJwIcumgbDdYiUT7zFQl3Wm1xD6fBKtah7H8ZJChj4wP+8C+QfeVy8xautR7rdmKEw==
  /glob/7.1.3:
    dependencies:
      fs.realpath: 1.0.0
      inflight: 1.0.6
      inherits: 2.0.3
      minimatch: 3.0.4
      once: 1.4.0
      path-is-absolute: 1.0.1
    dev: false
    resolution:
      integrity: sha512-vcfuiIxogLV4DlGBHIUOwI0IbrJ8HWPc4MU7HzviGeNho/UJDfi6B5p3sHeWIQ0KGIU0Jpxi5ZHxemQfLkkAwQ==
  /graceful-fs/4.1.15:
    dev: false
    resolution:
      integrity: sha512-6uHUhOPEBgQ24HM+r6b/QwWfZq+yiFcipKFrOFiBEnWdy5sdzYoi+pJeQaPI5qOLRFqWmAXUPQNsielzdLoecA==
  /inflight/1.0.6:
    dependencies:
      once: 1.4.0
      wrappy: 1.0.2
    dev: false
    resolution:
      integrity: sha1-Sb1jMdfQLQwJvJEKEHW6gWW1bfk=
  /inherits/2.0.3:
    dev: false
    resolution:
      integrity: sha1-Yzwsg+PaQqUC9SRmAiSA9CCCYd4=
  /iter-tools/6.1.6:
    dependencies:
      '@babel/runtime': 7.1.5
      dequeue: 1.0.5
      little-ds-toolkit: 1.1.0
      typescript-tuple: 1.7.3
      typescript-union: 0.0.0
    dev: false
    resolution:
      integrity: sha512-dCJ5FTc9xv10Cs84sd7oYt0mHbjcQQTKI+5K2E60qDVH5Jn7gtWhtRkeDeflIwRUC7W1sa+Jca/Vs5YT0B9DZw==
  /jsonfile/4.0.0:
    dev: false
    optionalDependencies:
      graceful-fs: 4.1.15
    resolution:
      integrity: sha1-h3Gq4HmbZAdrdmQPygWPnBDjPss=
  /little-ds-toolkit/1.1.0:
    dev: false
    resolution:
      integrity: sha512-2ARbmay5TrhFxFqcyFxPigmRToDuo8yxvIMdvwsDYNd/oZKnvsFT5L3I7s8WLXR7/tThYeffN9ouDtFFk4wzsA==
  /minimatch/3.0.4:
    dependencies:
      brace-expansion: 1.1.11
    dev: false
    resolution:
      integrity: sha512-yJHVQEhyqPLUTgt9B83PXu6W3rx4MvvHvSUvToogpwoGDOUQ+yDrR0HRot+yOCdCO7u4hX3pWft6kWBBcqh0UA==
  /once/1.4.0:
    dependencies:
      wrappy: 1.0.2
    dev: false
    resolution:
      integrity: sha1-WDsap3WWHUsROsF9nFC6753Xa9E=
  /path-is-absolute/1.0.1:
    dev: false
    engines:
      node: '>=0.10.0'
    resolution:
      integrity: sha1-F0uSaHNVNP+8es5r9TpanhtcX18=
  /regenerator-runtime/0.12.1:
    dev: false
    resolution:
      integrity: sha512-odxIc1/vDlo4iZcfXqRYFj0vpXFNoGdKMAUieAlFYO6m/nl5e9KR/beGf41z4a1FI+aQgtjhuaSlDxQ0hmkrHg==
  /semver/5.6.0:
    dev: false
    hasBin: true
    resolution:
      integrity: sha512-RS9R6R35NYgQn++fkDWaOmqGoj4Ek9gGs+DPxNUZKuwE183xjJroKvyo1IzVFeXvUrvmALy6FWD5xrdJT25gMg==
  /tslib/1.9.3:
    dev: false
    resolution:
      integrity: sha512-4krF8scpejhaOgqzBEcGM7yDIEfi0/8+8zDRZhNZZ2kjmHJ4hv3zCbQWxoJGz1iw5U0Jl0nma13xzHXcncMavQ==
  /typescript-miscellaneous/0.1.0:
    dependencies:
      '@types/node': 10.12.5
      tslib: 1.9.3
    dev: false
    engines:
      node: '>= 8.9.0'
      typescript: '>= 3.0.0'
    resolution:
      integrity: sha512-JPy2LNt44g/q4nJo4CzzGgfrA5LOHQl7I2XKVfK1gGz9F5WhataKRLrahk7RTMTckq31gkKe5LgiSkjxB2Ubiw==
  /typescript-tuple/1.7.3:
    dev: false
    engines:
      typescript: ^3.0.0
    resolution:
      integrity: sha512-ygCTfD6CX5ul+J0sZhq2MSouMNAeJpRXAFkYPUZomv457KWg15DuS7n+ubv15LtbX27iVo6DkuiX/5yyNdBvVg==
  /typescript-union/0.0.0:
    dependencies:
      typescript-tuple: 1.7.3
      utility-types: 2.1.0
    dev: false
    peerDependencies:
      typescript: '>= 3.0.0'
    resolution:
      integrity: sha512-sWOTNcTOXw27cJhfvjSbxrKggb4XD+fPIFkjIqqnycdVqStj/3IR3QVEMZOheWY5ELGS3PCwuHFDVWAbK9UHvA==
  /universalify/0.1.2:
    dev: false
    engines:
      node: '>= 4.0.0'
    resolution:
      integrity: sha512-rBJeI5CXAlmy1pV+617WB9J63U6XcazHHF2f2dbJix4XzpUF0RS3Zbj0FGIOCAva5P/d/GBOYaACQ1w+0azUkg==
  /utility-types/2.1.0:
    dev: false
    engines:
      node: '>= 4'
    resolution:
      integrity: sha512-/nP2gqavggo6l38rtQI/CdeV+2fmBGXVvHgj9kV2MAnms3TIi77Mz9BtapPFI0+GZQCqqom0vACQ+VlTTaCovw==
  /wrappy/1.0.2:
    dev: false
    resolution:
      integrity: sha1-tSQ9jz7BqjXxNkYFvA0QNuMKtp8=
registry: 'https://registry.npmjs.org/'
shrinkwrapMinorVersion: 9
shrinkwrapVersion: 3
specifiers:
  '@types/fs-extra': ^5.0.4
  '@types/node': ^10.12.0
  '@types/semver': ^5.5.0
  advanced-set-mutable-multi-key: ^0.0.1
  fs-extra: ^7.0.0
  glob: ^7.1.3
  glob-promise: ^3.4.0
  semver: ^5.6.0
  tslib: ^1.9.3
