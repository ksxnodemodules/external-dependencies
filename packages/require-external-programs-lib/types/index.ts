// tslint:disable:no-unnecessary-qualifier

import { KEYWORD, KEYNAME } from '../constants'
import { ManifestClassificationType } from '../enums'

export type ManifestList = ReadonlyArray<string>

export type ManifestClassification =
  ManifestClassification.Included |
  ManifestClassification.Excluded |
  ManifestClassification.Invalid

export namespace ManifestClassification {
  export interface Base {
    readonly type: ManifestClassificationType
  }

  export interface Included extends Base {
    readonly type: ManifestClassificationType.Included
  }

  export interface Excluded extends Base {
    readonly type: ManifestClassificationType.Excluded
  }

  export interface Invalid extends Base {
    readonly type: ManifestClassificationType.Invalid
    readonly reason: string
  }
}

export interface ManifestClassificationList {
  readonly included: ReadonlyArray<ManifestClassificationList.Included>
  readonly excluded: ReadonlyArray<ManifestClassificationList.Excluded>
  readonly invalid: ReadonlyArray<ManifestClassificationList.Invalid>
}

export namespace ManifestClassificationList {
  export interface Base {
    readonly manifestPath: string
    readonly manifestContent: ManifestContent
  }

  export interface Included extends Base {
    readonly manifestContent: ManifestContent.Included
  }

  export interface Excluded extends Base {}

  export interface Invalid extends Base {
    readonly reason: string
  }
}

export type ManifestContentList = ReadonlyArray<ManifestContentList.Item>

export namespace ManifestContentList {
  export interface Item {
    readonly manifestPath: string
    readonly manifestContent: ManifestContent
  }
}

export interface ManifestContent {
  readonly dependencies?: ManifestContent.DependencyDict
  readonly devDependencies?: ManifestContent.DependencyDict
  readonly peerDependencies?: {}
  readonly [KEYNAME]?: Requirements
}

export namespace ManifestContent {
  export type Included =
    Included.AsProd |
    Included.AsDev

  export namespace Included {
    export interface Base extends ManifestContent {
      readonly [KEYNAME]: Requirements
    }

    export interface AsProd extends Base {
      readonly dependencies: DependencyDict.Included
    }

    export interface AsDev extends Base {
      readonly devDependencies: DependencyDict.Included
    }
  }

  export interface DependencyDict {
    readonly [KEYWORD]?: string
  }

  export namespace DependencyDict {
    export interface Included extends DependencyDict {
      readonly [KEYWORD]: string
    }
  }
}

export type Requirements = ReadonlyArray<string>
