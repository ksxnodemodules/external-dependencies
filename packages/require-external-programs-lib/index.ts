export * from './types'
export * from './constants'
export * from './enums'
export * from './classes'
export * from './functions'
import { fromAllManifests } from './functions'
export default fromAllManifests
