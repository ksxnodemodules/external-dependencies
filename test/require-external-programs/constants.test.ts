import changeCase from 'change-case'
import { KEYWORD, KEYNAME } from 'require-external-programs-lib'

describe('require-external-programs-lib', () => {
  it('KEYWORD matches package name', () => {
    expect(KEYWORD).toBe('require-external-programs')
  })

  it('KEYNAME is lower camel case of KEYWORD', () => {
    expect(KEYNAME).toBe(changeCase.camel(KEYWORD))
  })
})
