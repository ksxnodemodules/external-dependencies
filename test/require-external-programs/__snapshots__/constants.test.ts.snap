// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`require-external-programs package.json has scripts.install 1`] = `"node ./install"`;
